cpp/src/strings/strings_column_view.cu

```diff
@@ -57,113 +55,37 @@ size_type strings_column_view::chars_size() const noexcept
-  CUDA_TRY(cudaMemcpy(
-    h_offsets.data(), d_output_offsets, (count + 1) * sizeof(size_type), cudaMemcpyDeviceToHost));
-  std::vector<char> h_buffer(buffer_size);
-  CUDA_TRY(cudaMemcpy(h_buffer.data(), d_buffer, buffer_size, cudaMemcpyDeviceToHost));
 
-  // print out the strings to stdout
-  for (size_type idx = 0; idx < count; ++idx) {
-    size_type offset = h_offsets[idx];
-    size_type length = h_offsets[idx + 1] - offset;
-    std::cout << idx << ":";
-    if (length)
-      std::cout << "[" << std::string(h_buffer.data() + offset) << "]";
-    else
-      std::cout << "<null>";
-    std::cout << delimiter;
-  }
-}
-
-//
-std::pair<rmm::device_vector<char>, rmm::device_vector<size_type>> create_offsets(
+std::pair<rmm::device_uvector<char>, rmm::device_uvector<size_type>> create_offsets(
   strings_column_view const& strings,
   rmm::cuda_stream_view stream,
   rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  size_type count          = strings.size();
-  const int32_t* d_offsets = strings.offsets().data<int32_t>();
+  size_type const count = strings.size();
+
+  auto d_offsets = strings.offsets().data<int32_t>();
   d_offsets += strings.offset();  // nvbug-2808421 : do not combine with the previous line
-  int32_t first = 0;
-  CUDA_TRY(
-    cudaMemcpyAsync(&first, d_offsets, sizeof(int32_t), cudaMemcpyDeviceToHost, stream.value()));
-  rmm::device_vector<size_type> offsets(count + 1);
+
+  rmm::device_uvector<size_type> offsets(count + 1, stream);
   // normalize the offset values for the column offset
-  thrust::transform(
-    rmm::exec_policy(stream),
-    d_offsets,
-    d_offsets + count + 1,
-    offsets.begin(),
-    [first] __device__(int32_t offset) { return static_cast<size_type>(offset - first); });
+  thrust::transform(rmm::exec_policy(stream),
+                    d_offsets,
+                    d_offsets + count + 1,
+                    offsets.begin(),
+                    [d_offsets] __device__(int32_t offset) {
+                      return static_cast<size_type>(offset - d_offsets[0]);
+                    });
   // copy the chars column data
-  int32_t bytes = 0;  // last offset entry is the size in bytes
-  CUDA_TRY(cudaMemcpyAsync(
-    &bytes, d_offsets + count, sizeof(int32_t), cudaMemcpyDeviceToHost, stream.value()));
+  auto const bytes = offsets.element(count, stream);
+  auto const chars_offset =
+    cudf::detail::get_value<offset_type>(strings.offsets(), strings.offset(), stream);
   stream.synchronize();
 
-  bytes -= first;
-  const char* d_chars = strings.chars().data<char>() + first;
-  rmm::device_vector<char> chars(bytes);
-  CUDA_TRY(
-    cudaMemcpyAsync(chars.data().get(), d_chars, bytes, cudaMemcpyDeviceToHost, stream.value()));
+  // bytes -= first;
+  const char* d_chars = strings.chars().data<char>() + chars_offset;
+  rmm::device_uvector<char> chars(bytes, stream);
+  CUDA_TRY(cudaMemcpyAsync(chars.data(), d_chars, bytes, cudaMemcpyDeviceToHost, stream.value()));
```

Comment:
This can't be right. This is a device-to-device copy, right?

```suggestion
  CUDA_TRY(cudaMemcpyAsync(chars.data(), d_chars, bytes, cudaMemcpyDefault, stream.value()));
```

Reply:
Oh, wow. This has been in there forever. I should've looked at it more closely. 
Do you think I should I use a `thrust::copy` here instead?

Reply:
Not yet. Thrust copy won't work unless we use only thrust vectors. Thrust::Async::copy might but I'm just starting to figure it out (undocumented).